src/Bicep.Types/Concrete/ResourceType.cs

```diff
@@ -11,27 +11,78 @@ namespace Azure.Bicep.Types.Concrete
+            string name, 
+            ScopeType? scopeType, 
+            ScopeType? readOnlyScopes, 
+            ITypeReference body, 
+            ResourceFlags? flags, 
+            IReadOnlyDictionary<string, ResourceTypeFunction>? functions, 
+            ScopeType? writableScopes = null, 
+            ScopeType? readableScopes = null)
+        {
+            // Check for illegal mixing of legacy and new scope fields
+            bool hasLegacy = scopeType.HasValue || readOnlyScopes.HasValue || (flags.HasValue && flags.Value != ResourceFlags.None);
+            bool hasModern = writableScopes.HasValue || readableScopes.HasValue;
 
-        public string Name { get; }
+            if (hasLegacy && hasModern)
+            {
+                throw new ArgumentException("Cannot mix both legacy scope fields (scopeType, readOnlyScopes, flags) and modern fields (writableScopes, readableScopes).");
+            }
 
-        public ScopeType ScopeType { get; }
+            Name = name;
+            Body = body;
+            Functions = functions;
 
-        public ScopeType? ReadOnlyScopes { get; }
+            // Legacy properties (kept for backward compatibility)
+#pragma warning disable CS0618 
+            this.ScopeType = scopeType;
+            this.ReadOnlyScopes = readOnlyScopes;
+            this.Flags = flags;
+#pragma warning restore CS0618
 
-        public ITypeReference Body { get; }
+            this.WritableScopes = writableScopes;
+            this.ReadableScopes = readableScopes;
+
+            if (!this.WritableScopes.HasValue && !scopeType.HasValue) {
+                throw new ArgumentException("Must supply either 'writableScopes' or 'scopeType'.");
+            }
+
+            if (!this.ReadableScopes.HasValue && !scopeType.HasValue) {
+                throw new ArgumentException("Must supply either 'readableScopes' or 'scopeType'.");
+            }
+        }
+
+        public string Name { get; }
 
-        public ResourceFlags Flags { get; }
+        public ITypeReference Body { get; }
 
         public IReadOnlyDictionary<string, ResourceTypeFunction>? Functions { get; }
+
+        [JsonIgnore(Condition = JsonIgnoreCondition.WhenWritingNull)]
+        public ScopeType? ReadableScopes { get; }
+
+        [JsonIgnore(Condition = JsonIgnoreCondition.WhenWritingNull)]
+        public ScopeType? WritableScopes { get; }
+
+        // Legacy properties (kept for backward compatibility)
+        [Obsolete("Use WritableScopes and ReadableScopes instead")]
+        [JsonIgnore(Condition = JsonIgnoreCondition.WhenWritingNull)]
+        public ScopeType? ScopeType { get; }
+
+        [Obsolete("Use ReadableScopes instead")]
+        [JsonIgnore(Condition = JsonIgnoreCondition.WhenWritingNull)]
+        public ScopeType? ReadOnlyScopes { get; }
+
+        [Obsolete("Use WritableScopes instead")]
+        [JsonIgnore(Condition = JsonIgnoreCondition.WhenWritingNull)]
+        public ResourceFlags? Flags { get; }
```

Comment:
Ideally, this class should be able to read in both serialized formats, but it should present a single way to read the values. That way, we can update consuming code to only care about the new fields while still supporting data that was serialized before the transition.

Does System.Text.Json still deserialize data as expected if `scopeType`, `readOnlyScopes`, and `flags` are constructor parameters but _aren't_ defined as class properties? Because the ideal would be for the class to look more like this:

```suggestion
        [JsonIgnore(Condition = JsonIgnoreCondition.WhenWritingNull)]
        public ScopeType ReadableScopes { get; }

        [JsonIgnore(Condition = JsonIgnoreCondition.WhenWritingNull)]
        public ScopeType WritableScopes { get; }
```

since `ReadableScopes` and `WritableScopes` can be derived from `ScopeType`, `ReadOnlyScopes`, and `Flags` (`ReadableScopes = ScopeType | ReadOnlyScopes; WritableScopes = Flags.HasFlag(ResourceFlags.ReadOnly) ? ScopeType.None : (ScopeType & ~ReadOnlyScopes);`